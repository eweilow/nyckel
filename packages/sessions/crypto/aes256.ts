// encrypt and decrypt are adapted from https://github.com/JamesMGreene/node-aes256/blob/master/index.js
// Original license:
/*
The MIT License (MIT)

Copyright (c) 2015 James M. Greene

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

import crypto from "crypto";

const rawTextEncoding = "utf8";
const encryptionEncoding = "base64";
const cipherType = "aes-256-ctr";

export function encrypt(key: string, plaintext: string) {
  if (typeof key !== "string" || !key) {
    throw new TypeError('Provided "key" must be a non-empty string');
  }
  if (typeof plaintext !== "string" || !plaintext) {
    throw new TypeError('Provided "plaintext" must be a non-empty string');
  }

  var sha256 = crypto.createHash("sha256");
  sha256.update(key);

  // Initialization Vector
  var iv = crypto.randomBytes(16);
  var cipher = crypto.createCipheriv(cipherType, sha256.digest(), iv);

  var ciphertext = cipher.update(Buffer.from(plaintext, rawTextEncoding));
  var encrypted = Buffer.concat([iv, ciphertext, cipher.final()]).toString(
    encryptionEncoding
  );

  return encrypted;
}

export function decrypt(key: string, encrypted: string) {
  if (typeof key !== "string" || !key) {
    throw new TypeError('Provided "key" must be a non-empty string');
  }
  if (typeof encrypted !== "string" || !encrypted) {
    throw new TypeError('Provided "encrypted" must be a non-empty string');
  }

  var sha256 = crypto.createHash("sha256");
  sha256.update(key);

  var input = Buffer.from(encrypted, "base64");

  if (input.length < 17) {
    throw new TypeError(
      'Provided "encrypted" must decrypt to a non-empty string'
    );
  }

  var iv = input.slice(0, 16);
  var decipher = crypto.createDecipheriv(cipherType, sha256.digest(), iv);

  var ciphertext = input.slice(16);
  var plaintext = Buffer.concat([
    decipher.update(ciphertext),
    decipher.final()
  ]).toString(rawTextEncoding);

  return plaintext;
}
