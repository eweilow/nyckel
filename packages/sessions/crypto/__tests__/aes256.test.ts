import { encrypt, decrypt } from "../aes256";

const invalidCases = [undefined, null, "", 0, true, () => {}, new Date()];

describe("aes256", () => {
  describe("encrypt", () => {
    it("throws if key not provided properly", () => {
      for (const c of invalidCases) {
        expect(() => {
          encrypt(c as any, "plaintext");
        }).toThrowErrorMatchingSnapshot();
      }
    });
    it("throws if plaintext not provided properly", () => {
      for (const c of invalidCases) {
        expect(() => {
          encrypt("key", c as any);
        }).toThrowErrorMatchingSnapshot();
      }
    });
  });

  describe("decrypt", () => {
    it("throws if key not provided properly", () => {
      for (const c of invalidCases) {
        expect(() => {
          decrypt(c as any, "encrypted");
        }).toThrowErrorMatchingSnapshot();
      }
    });

    it("throws if encrypted not provided properly", () => {
      for (const c of invalidCases) {
        expect(() => {
          decrypt("key", c as any);
        }).toThrowErrorMatchingSnapshot();
      }
    });

    it("throws if input length is less than 17", () => {
      const input = "test";
      expect(() => {
        decrypt("key", input);
      }).toThrowErrorMatchingSnapshot();
    });
  });

  it("encrypts and decrypts correctly", () => {
    const key = "key";
    const value = "value";
    const encrypted = encrypt(key, value);
    expect(decrypt(key, encrypted)).toBe(value);
  });

  it("randomly initializes first 16 bytes", () => {
    const key = "key";
    const value = "value";
    const encrypted1 = encrypt(key, value);
    const encrypted2 = encrypt(key, value);
    expect(encrypted1).not.toEqual(encrypted2);

    const decrypt1 = decrypt(key, encrypted1);
    const decrypt2 = decrypt(key, encrypted2);
    expect(decrypt1).toEqual(decrypt2);
  });
});
