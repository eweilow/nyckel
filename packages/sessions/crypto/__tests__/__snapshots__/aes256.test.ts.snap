// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aes256 decrypt throws if encrypted not provided properly 1`] = `"Provided 'encrypted' must be a non-empty string"`;

exports[`aes256 decrypt throws if encrypted not provided properly 2`] = `"Provided 'encrypted' must be a non-empty string"`;

exports[`aes256 decrypt throws if encrypted not provided properly 3`] = `"Provided 'encrypted' must be a non-empty string"`;

exports[`aes256 decrypt throws if encrypted not provided properly 4`] = `"Provided 'encrypted' must be a non-empty string"`;

exports[`aes256 decrypt throws if encrypted not provided properly 5`] = `"Provided 'encrypted' must be a non-empty string"`;

exports[`aes256 decrypt throws if encrypted not provided properly 6`] = `"Provided 'encrypted' must be a non-empty string"`;

exports[`aes256 decrypt throws if encrypted not provided properly 7`] = `"Provided 'encrypted' must be a non-empty string"`;

exports[`aes256 decrypt throws if input length is less than 17 1`] = `"Provided \\"encrypted\\" must decrypt to a non-empty string"`;

exports[`aes256 decrypt throws if key not provided properly 1`] = `"Provided 'key' must be a non-empty string"`;

exports[`aes256 decrypt throws if key not provided properly 2`] = `"Provided 'key' must be a non-empty string"`;

exports[`aes256 decrypt throws if key not provided properly 3`] = `"Provided 'key' must be a non-empty string"`;

exports[`aes256 decrypt throws if key not provided properly 4`] = `"Provided 'key' must be a non-empty string"`;

exports[`aes256 decrypt throws if key not provided properly 5`] = `"Provided 'key' must be a non-empty string"`;

exports[`aes256 decrypt throws if key not provided properly 6`] = `"Provided 'key' must be a non-empty string"`;

exports[`aes256 decrypt throws if key not provided properly 7`] = `"Provided 'key' must be a non-empty string"`;

exports[`aes256 encrypt throws if key not provided properly 1`] = `"Provided 'key' must be a non-empty string"`;

exports[`aes256 encrypt throws if key not provided properly 2`] = `"Provided 'key' must be a non-empty string"`;

exports[`aes256 encrypt throws if key not provided properly 3`] = `"Provided 'key' must be a non-empty string"`;

exports[`aes256 encrypt throws if key not provided properly 4`] = `"Provided 'key' must be a non-empty string"`;

exports[`aes256 encrypt throws if key not provided properly 5`] = `"Provided 'key' must be a non-empty string"`;

exports[`aes256 encrypt throws if key not provided properly 6`] = `"Provided 'key' must be a non-empty string"`;

exports[`aes256 encrypt throws if key not provided properly 7`] = `"Provided 'key' must be a non-empty string"`;

exports[`aes256 encrypt throws if plaintext not provided properly 1`] = `"Provided 'plaintext' must be a non-empty string"`;

exports[`aes256 encrypt throws if plaintext not provided properly 2`] = `"Provided 'plaintext' must be a non-empty string"`;

exports[`aes256 encrypt throws if plaintext not provided properly 3`] = `"Provided 'plaintext' must be a non-empty string"`;

exports[`aes256 encrypt throws if plaintext not provided properly 4`] = `"Provided 'plaintext' must be a non-empty string"`;

exports[`aes256 encrypt throws if plaintext not provided properly 5`] = `"Provided 'plaintext' must be a non-empty string"`;

exports[`aes256 encrypt throws if plaintext not provided properly 6`] = `"Provided 'plaintext' must be a non-empty string"`;

exports[`aes256 encrypt throws if plaintext not provided properly 7`] = `"Provided 'plaintext' must be a non-empty string"`;
